[X86] Recognise VPMADD52L pattern with AVX512IFMA/AVXIFMA (#153787)

Arghnews opened the pull request:
Match `(X * Y) + Z` in `combineAdd`. If target supports and we don't overflow (ie. we know the top 12 bits are unset), rewrite using VPMADD52L

Have deliberately not put tests into existing `combine-vpmadd52.ll` as the flags make it a pain

Have just done the `L` version for now at least, wanted to get feedback before continuing

@RKSimon

Arghnews committed:
[X86] Recognise VPMADD52L pattern with AVX512IFMA/AVXIFMA (#153787)

Match (X * Y) + Z in combineAdd. If target supports and we don't
overflow, rewrite using VPMADD52L

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -57966,6 +57966,51 @@ static SDValue pushAddIntoCmovOfConsts(SDNode *N, const SDLoc &DL,
                      Cmov.getOperand(3));
 }
 
+static SDValue matchIntegerMultiplyAdd(SDNode *N, SelectionDAG &DAG,
```
matchIntegerMultiplyAdd -> matchVPMADD52

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -57966,6 +57966,51 @@ static SDValue pushAddIntoCmovOfConsts(SDNode *N, const SDLoc &DL,
 }
 
+static SDValue matchIntegerMultiplyAdd(SDNode *N, SelectionDAG &DAG,
+                                       SDValue Op0, SDValue Op1,
```
don't bother passing Op0/Op1 - we get them as X and Y from the sdk_match

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -57966,6 +57966,51 @@ static SDValue pushAddIntoCmovOfConsts(SDNode *N, const SDLoc &DL,
+  if (!sd_match(N, m_Add(m_Mul(m_Value(X), m_Value(Y)), m_Value(Acc))))
+    return SDValue();
+
+  auto CheckMulOperand = [&DAG, &VT](const SDValue &M, SDValue &Xval,
```
Do you need all this? Surely you just need to check that the knownbits of X, Y and KnownBits::mul(X,y) are all known zero in the upper 12 bits?

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -58069,6 +58114,11 @@ static SDValue combineAdd(SDNode *N, SelectionDAG &DAG,
   }
 
+  if (SDValue node =
+          matchIntegerMultiplyAdd(N, DAG, Op0, Op1, DL, VT, Subtarget)) {
```
(style) node -> IFMA52

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -58069,6 +58114,11 @@ static SDValue combineAdd(SDNode *N, SelectionDAG &DAG,
+  if (SDValue node =
+          matchIntegerMultiplyAdd(N, DAG, Op0, Op1, DL, VT, Subtarget)) {
+    return node;
+  }
```
(style) drop unnecessary braces

Arghnews committed:
Apply review feedback, simplify, add more tests

Arghnews committed:
Refactor test, remove deprecated undef

Arghnews committed:
Implement SplitOpsAndApply

Arghnews commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -57966,6 +57966,51 @@ static SDValue pushAddIntoCmovOfConsts(SDNode *N, const SDLoc &DL,
+  if (!VT.isVector() || VT.getScalarType() != MVT::i64 ||
+      !Subtarget.hasAVX512() ||
+      (!Subtarget.hasAVXIFMA() && !Subtarget.hasIFMA()) ||
+      !DAG.getTargetLoweringInfo().isOperationLegalOrCustom(X86ISD::VPMADD52L,
```
Believe I've done this now. Put in a test (last one in the file) using `<16 x i64>` that fires to try and confirm I've got it right

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -27,6 +27,7 @@
 #include "llvm/ADT/StringSwitch.h"
 #include "llvm/Analysis/BlockFrequencyInfo.h"
 #include "llvm/Analysis/ProfileSummaryInfo.h"
+#include "llvm/Analysis/ValueTracking.h"
```
remove this

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -4461,6 +4462,7 @@ SDValue SplitOpsAndApply(SelectionDAG &DAG, const X86Subtarget &Subtarget,
   unsigned NumSubs = 1;
   if ((CheckBWI && Subtarget.useBWIRegs()) ||
       (!CheckBWI && Subtarget.useAVX512Regs())) {
+    // if (0) {
```
remove this

RKSimon commented on llvm/test/CodeGen/X86/ifma-combine-vpmadd52.ll:
```diff
@@ -0,0 +1,217 @@
+; 4503599627370496 == (1 << 52)
+; 4503599627370495 == (1 << 52) - 1
+
+define dso_local <8 x i64> @test_512_combine_evex(<8 x i64> noundef %0, <8 x i64> noundef %1, <8 x i64> noundef %2) local_unnamed_addr #0 {
```
remove all uses of dso_local / noundef / local_unnamed_addr

RKSimon commented on llvm/test/CodeGen/X86/ifma-combine-vpmadd52.ll:
```diff
@@ -0,0 +1,217 @@
+; RUN: llc < %s -O1 -mtriple=x86_64-unknown-unknown | FileCheck %s --check-prefixes=X64
```
don't use attributes (which should allow you to remove alot of duplicate tests below) - we need to use a selection of RUN lines:
```
; RUN: llc < %s -mtriple=x86_64-- -mattr=+avxifma | FileCheck %s --check-prefixes=X64,AVX
; RUN: llc < %s -mtriple=x86_64-- -mattr=+avx512ifma | FileCheck %s --check-prefixes=X64,AVX512,AVX512-NOVL
; RUN: llc < %s -mtriple=x86_64-- -mattr=+avx512ifma,+avx512vl | FileCheck %s --check-prefixes=X64,AVX512,AVX512VL
```

Arghnews replied:
Have made this change. Apologies, I did check other tests and many of them use the attributes lines so I presumed it was acceptable. It is much neater now.

I wasn't sure exactly the combination of attributes that needed testing, appreciate you providing it here

RKSimon commented on llvm/test/CodeGen/X86/ifma-combine-vpmadd52.ll:
```diff
@@ -0,0 +1,217 @@
+; X64:       retq
+  %4 = and <8 x i64> %0, splat (i64 4503599627370495)
+  %5 = and <8 x i64> %1, splat (i64 4503599627370495)
+  %6 = and <8 x i64> %2, splat (i64 4503599627370495)
```
the accumulator doesn't need to be masked

XChy commented on llvm/test/CodeGen/X86/ifma-combine-vpmadd52.ll:
```diff
@@ -0,0 +1,217 @@
+; X64:       ret
+entry:
+  ; product = 2^52
+  %mul = mul <8 x i64> splat(i64 2251799813685248), splat(i64 2)
```
`%mul` constant folds early. It has no impact on this patch.

XChy commented on llvm/test/CodeGen/X86/ifma-combine-vpmadd52.ll:
```diff
@@ -0,0 +1,217 @@
+; X64-NOT:     vpadd
+; X64-NEXT:    {vex} vpmadd52luq %ymm1, %ymm2, %ymm0
+; X64-NEXT:    retq
+  %4 = and <4 x i64> %0, <i64 67108863, i64 67108863, i64 67108863, i64 67108863>
```
`splat (i64 67108863)` is cleaner.

XChy commented on llvm/test/CodeGen/X86/ifma-combine-vpmadd52.ll:
```diff
@@ -0,0 +1,217 @@
+; X64-NOT:     vpadd
+; X64-NEXT:    vpmadd52luq %ymm1, %ymm2, %ymm0
+; X64-NEXT:    retq
+  %4 = and <4 x i64> %0, <i64 67108863, i64 67108863, i64 67108863, i64 67108863>
```
splat (i64 67108863) is cleaner.

XChy reviewed (commented):
Please split the patch into two commits. The first commit is to add the test cases. The second adds the code change and testcase diff.

Arghnews committed:
Refactor and cleanup tests

Changed all constants to splats
Removed const_2pow51_times_2 as folds early
Remove unnecessary acc masks
Change numbered vars to more readable names
Remove attributes and use RUN lines
Remove dso_local/noundef/local_unnamed_addr

Arghnews committed:
Refactor X86ISelLowering.cpp and add test checks

Address feedback:
Removed ValueTracking.h (sorry think clion autoadded and I missed it)
Removed leftover comment // if (0
Added !Subtarget.hasAVXIFMA guard to prevent missed application of this
change on AVXIFMA 128/256bit
Add assert message
Change X86ISD::VPMADD52L call to reflect different X86 op order
Add test CHECK lines from script

Arghnews commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -57966,6 +57968,43 @@ static SDValue pushAddIntoCmovOfConsts(SDNode *N, const SDLoc &DL,
+    return SDValue();
+
+  // Need AVX-512VL vector length extensions if operating on XMM/YMM registers
+  if (!Subtarget.hasVLX() && VT.getSizeInBits() < 512)
```
Thanks for spotting, have added another condition now to keep this transform on those targets (and tests now fail if I remove the additional if clause)

Arghnews commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -57966,6 +57968,43 @@ static SDValue pushAddIntoCmovOfConsts(SDNode *N, const SDLoc &DL,
+    EVT SubVT = SubOps[0].getValueType();
+    assert(SubVT.getScalarSizeInBits() == 64);
+    return G.getNode(X86ISD::VPMADD52L, DL, SubVT, SubOps[0] /*Acc*/,
+                     SubOps[1] /*X*/, SubOps[2] /*Y*/);
```
Thanks, I missed this, have changed op order now

RKSimon commented on llvm/test/CodeGen/X86/ifma-combine-vpmadd52.ll:
```diff
@@ -0,0 +1,357 @@
+  ret <8 x i64> %res
+}
+
+define <16 x i64> @test_1024_combine_split(<16 x i64> %x, <16 x i64> %y, <16 x i64> %z) {
```
Use nounwind to silence the cfi noise
`define <16 x i64> @test_1024_combine_split(<16 x i64> %x, <16 x i64> %y, <16 x i64> %z) nounwind {`

Arghnews replied:
Done

Arghnews committed:
Add nounwind to silence cfi noise

XChy reviewed (approved):
LGTM, cheers. Please wait for @RKSimon to confirm.
It would be better if the unoptimized testcases are updated in the first commit. For example, https://github.com/llvm/llvm-project/pull/156847/commits/f4deffa2fb70968adcce213c0a9f16f11d00b756. The diff makes review easier. Just a reminder for the follow-up patches.

RKSimon reviewed (changes requested):
Please can you add a couple of tests for odd types that satisfy the knownbits (<1 x i64> and <6 x i64> might be good examples)

Arghnews committed:
Add tests for odd sizes

Fix variable case

RKSimon reviewed (commented):
A few minors but it looks like its almost there

Arghnews committed:
Addressed feedback

Fixed testcase return, fixed naming, moved comment
Early out on known bits

RKSimon reviewed (approved):
LGTM - cheers

RKSimon committed:
Merge branch 'main' into VPMADD52_ifma_combine
